### ISSUE TITLE
Make it possible to edit diff

### ISSUE THREAD

**Issue by tarsius** [magit/magit#2938]
Sometimes it would be nice to edit a diff right in the status or diff buffer (when showing unstaged or staged changes).

Emacs' `diff-mode` supports this to some extend.  `diff-mode` buffers are editable, so if such a buffer is visiting an actual `*.patch` file, then you can just save the buffer and that file is updated.  One nice thing about this feature is that line numbers in hunk headers are automatically adjusted, so `patch` will later not complain about invalid line-numbers. (Magit does the same line-number adjustment when applying parts of a hunk, this isn't the hard part.)

But when the `diff-mode` buffer is not visiting a file, then that doesn't work.  You can still edit the diff and then save it to some new file, I suppose, but that's usually not very useful.

The diffs shown by Magit never visit a file.  (Magit has some very rudimentary support for patches received by mail, but if you visit one of those, then Magit actually uses `diff-mode` for that purpose!)  Unstaged changes, for example, are the output of the command `git diff`.  To "save" changes made to that diff as shown in some Magit buffer Magit would have to edit one or more "content files" (not patch files).

That's doable but non-trivial.  However I intend to implement it eventually, just consider this scenario: 

- The same line contains unrelated changes.  E.g. you made some changes to a line, like renaming a function, and also removed trailing whitespace because you happened to notice that there was some.  Now you want to commit these two changes using separate commits.

  Currently you would have to go the file, undo one of the changes and stage, then go to the file again and redo the change.

  Once it is possible to edit diffs, you could stage both changes and then edit the staged changes to remove one of them.  There would be no need to go to the file twice, once to undo a change and then again to redo it.

This feature will probably work something like this:

1. The user explicitly enters "edit mode", which overrides some Magit keybindings, so that e.g. <kbd>s</kbd> enters the character "s" instead of staging the current hunk.
2. The user then makes the changes.
3. And finally exits "edit mode".  At that time Magit applies the changes to the appropriate files in the working tree (or in the example above to the index). 

As a side-note, I plan to generalize `magit-diff-mode` so that it can be used for everything `diff-mode` can be used for. (But of course with all the magit-specific goodies.)

**Comment by kyleam** [comment]
Related to #2718 and #2173.

**Comment by vermiculus** [comment]
Cross referencing a related conversation on Emacs SE: https://emacs.stackexchange.com/q/35960/2264

**Comment by fice-t** [comment]
It would also be useful to have a corresponding "edit mode" for revision buffers where one is able to edit the diff, and also the commit message, notes, author/commit info, and so on.

**Comment by tarsius** [comment]
It's unlikely that I will implement that any time soon.

**Comment by fice-t** [comment]
@tarsius I understand putting this issue off, but I'm not sure why the issue should be closed (or labeled `wontadd`). If it's because it's only a `wontadd` for Year of Magit, then what about leaving it open and creating a label like `After Year of Magit` (if only to prevent people from creating duplicate issues about this feature)?

**Comment by tarsius** [comment]
I occasionally go back and look at `wontadd` issues, so there is hope. But I currently have no plans of adding this so I closed it. I will deal with duplicates as they happen.

**Comment by jumarko** [comment]
This would be a cool feature! Quick fixes could be made much faster without having to jump back to source buffers.

**Comment by kees-jan** [comment]
I, too, would love to see this feature added. I guess is this is one of those things you don't know you're missing until you see someone else use it.

In my case: CLion supports this, and it is awesome ;-)

**Comment by thisch** [comment]
@kees-jan Do you have a link to the release-notes or the documentation of this feature of the JetBrains IDEs? I'm curious how their user-interface for this feature looks like.

**Comment by kees-jan** [comment]
I could not find a very detailed explanation. [This link](https://www.jetbrains.com/help/clion/differences-viewer.html#diff-merge-viewer) mentions that you can edit code in the viewer, but doesn't say much more than that. 

If it helps, I prepared the screenshot below. On the left the unstaged changes (checkbox to indicate that they will be staged and committed). There, I can right click and choose "show diff"

On the right the differences viewer (currently in "unified mode". It behaves like a regular editor, in that I can click anywhere and start typing (well, except on the deleted lines). When I do, the "patch" is updated as one would expect (up to the point that more context lines are added) and the underlying file is updated immediately.

![image](https://user-images.githubusercontent.com/221402/148197763-7c205efb-b104-4d82-aecd-6c8e29b179ce.png)

**Comment by ChanderG** [comment]
Had a similar need today - to make the same fix in multiple files, already committed into HEAD.

I kinda simulated this workflow (for editing HEAD) by doing the following:
1. `W c` - create a patch from the needed commit.
2. Open the patch, edit it as needed (`diff-mode` allows this) and save it.
3. (Hard) Reset HEAD.
4. `w w` - apply the patch.
5. (optional) `c w` to reword the commit msg.

Obviously, if I needed to edit an older commit, would have been a bit more involved - in interactive rebase with a break in the middle - where you manually apply the patch.

I know this is reaching the edges of the git model (and going into patch based VC territory), but just thinking about some options out loud for fun:
1. An option in reset `X` to produce a patch of the commit (or staged/unstaged changes).
2. A command in rebase `r` (like `r m` modify and `r w` reword) to edit a particular commit as a patch - basically automate the above manual process.
3. I see that we can already convert a stash into a patch using `z f`. So maybe similar commands for unstaged/staged changes?

**Comment by thisch** [comment]
Thx a lot for the comment @ChanderG! I just had a look again at the issue and I think we can create a simple elisp command that basically
* creates the patch (`W c c`) 
* opens the patch via `view-file`

IMO it would be nice if in the opened file `C-c C-c` 

* saved the file
* reset --hard one commit
* applied the patch

**Comment by jflatow** [comment]
This would be very helpful when working with large diffs a la `wdired`/`wgrep`. @tarsius if you got some sponsors would you reconsider the `wont add` labe?

**Comment by tarsius** [comment]
@jflatow I have to prioritize and manage the time available to me. I would like to make substantial improvements to how diffing works. These changes have substantial prerequisites, that are multi-month projects in their own right.

I could instead implement "edit diff" in a more hacky way. That would mean that I would have to invest time on that hacky implementation now, and then when I implement the planned improvements to diffs and sections in general, that work would be slowed down by having to carry that hacky implementation along. Then finally, when the prerequisites for doing a proper implementation are in place, I would have to invest time again to do that.

**Comment by jflatow** [comment]
@tarsius then it sounds like you are actually planning to address eventually, just in a proper non-hacky way, so `wont add` is then maybe not the appropriate label 🙏 😅

**Comment by vermiculus** [comment]
For anyone else who happens across this thread as I did:

I was getting pretty jealous of my fugitive.vim friends' ability to edit files in the index, so I decided to start hacking away at this. At least perhaps I'd have some magit extension package to publish by the end of it. As I kept digging, I discovered 'pseudo-revs' like `{worktree}` and `{index}` and eventually found my way to `magit-find-file-index-noselect` and `magit-update-index` – which led me to 648540a6c6bce898b5313a88e348026c40884554.

Little did *I* know that magit's ediff integration already supports this! It works flawlessly for my needs, at least. It's not exactly what's described in roadmap post at the top of this thread, but it gets the job done.

More than ten years using and hacking at this tool and I'm still learning new things.

Time for me to learn ediff :-)
